src/Listener.cc

```diff
@@ -20,13 +20,21 @@
 
 using namespace std;
 
+thread_local bool CBLListenerToken::_inCallback = false;
 
 void CBLListenerToken::remove() {
     auto oldOwner = _owner;
     if (oldOwner) {
         _callback = nullptr;
         _owner = nullptr;
         oldOwner->remove(this);
+
+        // If the listener is removed from within its callback, don't wait for it to finish.
+        // Otherwise the callback would deadlock waiting for itself to finish.
```

Comment:
👍 Since we have a mutex in `CBLListenerToken` now anyway, I've put access to `_callback` under the mutex instead of storing  `_callback` in an atomic.